stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -715,416 +796,440 @@ for (t, unitt) in ((UpperTriangular, UnitUpperTriangular),
         throw(DimensionMismatch("right hand side B needs first dimension of size $(size(A,1)), has size $m"))
     end
+    Adata = triangulardata(A)
     for j = 1:n
         for i = 1:m
             Bij = B[i,j]
             for k = i + 1:m
-                Bij += A.data[k,i]'B[k,j]
+                Bij += Adata[k,i]'B[k,j]
             end
             B[i,j] = Bij
         end
     end
     B
 end
 
-function lmul!(transA::Transpose{<:Any,<:UpperTriangular}, B::StridedVecOrMat)
-    A = transA.parent
+function _lmul!(transA::AbstractMatrix, B::AbstractVecOrMat, ::LowerTriangularLayout{<:AbstractRowMajor}, _)
+    A = transpose(transA)
     m, n = size(B, 1), size(B, 2)
     if m != size(A, 1)
         throw(DimensionMismatch("right hand side B needs first dimension of size $(size(A,1)), has size $m"))
     end
+    Adata = triangulardata(A)
     for j = 1:n
         for i = m:-1:1
-            Bij = transpose(A.data[i,i]) * B[i,j]
+            Bij = transpose(Adata[i,i]) * B[i,j]
             for k = 1:i - 1
-                Bij += transpose(A.data[k,i]) * B[k,j]
+                Bij += transpose(Adata[k,i]) * B[k,j]
             end
             B[i,j] = Bij
         end
     end
     B
 end
-function lmul!(transA::Transpose{<:Any,<:UnitUpperTriangular}, B::StridedVecOrMat)
-    A = transA.parent
+function _lmul!(transA::AbstractMatrix, B::AbstractVecOrMat, ::UnitLowerTriangularLayout{<:AbstractRowMajor}, _)
+    A = transpose(transA)
     m, n = size(B, 1), size(B, 2)
     if m != size(A, 1)
         throw(DimensionMismatch("right hand side B needs first dimension of size $(size(A,1)), has size $m"))
     end
+    Adata = triangulardata(A)
     for j = 1:n
         for i = m:-1:1
             Bij = B[i,j]
             for k = 1:i - 1
-                Bij += transpose(A.data[k,i]) * B[k,j]
+                Bij += transpose(Adata[k,i]) * B[k,j]
             end
             B[i,j] = Bij
         end
     end
     B
 end
 
-function lmul!(transA::Transpose{<:Any,<:LowerTriangular}, B::StridedVecOrMat)
-    A = transA.parent
+function _lmul!(transA::AbstractMatrix, B::AbstractVecOrMat, ::UpperTriangularLayout{<:AbstractRowMajor}, _)
+    A = transpose(transA)
     m, n = size(B, 1), size(B, 2)
     if m != size(A, 1)
         throw(DimensionMismatch("right hand side B needs first dimension of size $(size(A,1)), has size $m"))
     end
+    Adata = triangulardata(A)
     for j = 1:n
         for i = 1:m
-            Bij = transpose(A.data[i,i]) * B[i,j]
+            Bij = transpose(Adata[i,i]) * B[i,j]
             for k = i + 1:m
-                Bij += transpose(A.data[k,i]) * B[k,j]
+                Bij += transpose(Adata[k,i]) * B[k,j]
             end
             B[i,j] = Bij
         end
     end
     B
 end
-function lmul!(transA::Transpose{<:Any,<:UnitLowerTriangular}, B::StridedVecOrMat)
-    A = transA.parent
+function _lmul!(transA::AbstractMatrix, B::AbstractVecOrMat, ::UnitUpperTriangularLayout{<:AbstractRowMajor}, _)
```

Comment:
Here you're using the stridedness as a more generic "optimal order of iteration" kind of thing — whereas before these methods would work for generic (non-strided) triangular data backing arrays.

`Adata = triangulardata(A)` also feels weird to me — now not only is the stridedness determining how elements are laid out in memory, it's also determining if `Adata[i,j]` is really `A[j, i]`?

Reply:
It feels weird to me too, but it was necessary to get for example
```julia
view(UpperTriangular(A), Base.OneTo(3), Base.OneTo(3))
```
to work with the same implementations.

Ideally, this could change to an `@inbounds`-like macro for all sparse matrix types, something like `@insparse A[k,j]` would lower to `@inbounds A.data[k,j]`. Unfortunately it does not look possible to make your own macros like this.

Reply:
There are two highly coupled but very distinct APIs here:

1. How do you load an element based upon an array's _pointer_ and in-memory strides?
2. How do you transform a set of indices to load an element based upon an array's _parent_ and Julian indexing into it?

They happen to coincide for views into `Array`.  We can also easily recompute accesses into views of _any_ Julian array if the indices in the view form a strided pattern.  Tackling \#1 is a lot easier than \#2… and for the most part we've managed to get the optimizations we need for \#2 purely based upon `IndexLinear` — in terms of a `StridedArray`, an `IndexLinear` array is one that has a fixed offset between all elements.  Once we lose that, though, views and wrappers become totally opaque black boxes, whereas strided arrays are still able to introspect and optimize the root access pattern.

Crazy half-baked idea that I'm absolutely not suggesting as a change for this PR but is worth considering: what would `IndexStrided <: IndexingStyle` look like?

Reply:
I think your description of the distinction between (1) and (2) is a bit inaccurate: when `T` is not a `BlasFloat`, `pointer` is never called, but `MemoryLayout` still dictates how memory is accessed. 

Merging with `IndexStyling` is a good idea: I'm pretty sure one can always infer whether its `IndexCartesian` or `IndexLinear` from the memory layout.

> views and wrappers become totally opaque black boxes

I don't think this is right, and again, the main motivation of this PR is views of banded matrices, which are banded (as far as BLAS is concerned).